File: problems/22-subsequence-dp/001-longest-increasing-subsequence/java/Solution.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
    public int lengthOfLIS(int[] nums) {
        List<Integer> tails = new ArrayList<>();

        for (int value : nums) {
            int left = 0;
            int right = tails.size();

            while (left < right) {
                int mid = left + (right - left) / 2;
                if (tails.get(mid) < value) {
                    left = mid + 1;
                } else {
                    right = mid;
                }
            }

            if (left == tails.size()) {
                tails.add(value);
            } else {
                tails.set(left, value);
            }
        }

        return tails.size();
    }
}

File: problems/22-subsequence-dp/001-longest-increasing-subsequence/python/solution.py
from bisect import bisect_left


class Solution:
    def lengthOfLIS(self, nums: list[int]) -> int:
        tails = []

        for value in nums:
            index = bisect_left(tails, value)
            if index == len(tails):
                tails.append(value)
            else:
                tails[index] = value

        return len(tails)
